File: leetcode/easy/604-design-compressed-string-iterator/604-design-compressed-string-iterator.ts
// 604-design-compressed-string-iterator
// URL: https://leetcode.com/problems/design-compressed-string-iterator/description/?envType=study-plan-v2&envId=premium-algo-100


class StringIterator {
    private complessedList: [string, number][] = [];
    currentChar: string = '';
    currentCount: number = 0;
  
    constructor(compressedString: string) {
      this.complessedList = this.parseString(compressedString);
    }
  
    parseString(str: string): [string, number][] {
      const regex = /\d+|[a-zA-Z]+/g;
      const matches = str.match(regex) || [];
      const result: [string, number][] = [];
  
      for (let i = 0; i < matches.length; i += 2) {
        const word = matches[i];
        const number = matches[i + 1] ? parseInt(matches[i + 1], 10) : NaN;
        if (!isNaN(number)) {
          result.push([word, number]);
        }
      }
  
      return result;
    }
  
    moveToNextComplessed() {
      this.currentChar = this.complessedList[0][0];
      this.currentCount = this.complessedList[0][1];
      this.complessedList.shift();
    }
  
    next(): string {
      if (this.currentCount === 0) {
        if (this.complessedList.length === 0) {
          return ' ';
        }
        this.moveToNextComplessed();
      }
      this.currentCount -= 1;
  
      return this.currentChar;
    }
  
    hasNext(): boolean {
      return this.complessedList.length > 0 || this.currentCount > 0;
    }
  }
  
  /**
   * Your StringIterator object will be instantiated and called as such:
   * var obj = new StringIterator(compressedString)
   * var param_1 = obj.next()
   * var param_2 = obj.hasNext()
   */
  

File: leetcode/easy/604-design-compressed-string-iterator/604-design-compressed-string-iterator.py
import re


class StringIterator:

    def __init__(self, compressedString: str):
        self.currentChar = ''
        self.currentCount = 0
        self.compressedList = StringIterator.parse_string(compressedString)

    def parse_string(string):
        regex = r'\d+|[a-zA-Z]'
        matches = re.findall(regex, string)
        result = []

        for i in range(0, len(matches), 2):
            word = matches[i]
            number_str = matches[i + 1] if i + 1 < len(matches) else None
            if number_str:
                number = int(number_str)
            result.append([word, number])

        return result

    def moveToNext(self):
        self.currentChar, self.currentCount = self.compressedList.pop(0)

    def next(self) -> str:
        if not self.hasNext():
            return ' '
        if self.currentCount == 0:
            self.moveToNext()

        self.currentCount -= 1

        print(self.currentChar, self.currentCount)
        print(self.compressedList)

        return self.currentChar

    def hasNext(self) -> bool:
        if self.currentCount == 0 and len(self.compressedList) == 0:
            return False

        return True

# Your StringIterator object will be instantiated and called as such:
# obj = StringIterator(compressedString)
# param_1 = obj.next()
# param_2 = obj.hasNext()
